pkg/controller/accountclaim/accountclaim_finalizer_test.go

```diff
@@ -0,0 +1,179 @@
+package accountclaim
+
+import (
+	"fmt"
+
+	"github.com/golang/mock/gomock"
+	"github.com/openshift/aws-account-operator/pkg/apis"
+	"github.com/openshift/aws-account-operator/pkg/apis/aws/v1alpha1"
+	awsv1alpha1 "github.com/openshift/aws-account-operator/pkg/apis/aws/v1alpha1"
+	"github.com/openshift/aws-account-operator/pkg/awsclient/mock"
+	"github.com/openshift/aws-account-operator/pkg/controller/testutils"
+	"github.com/openshift/aws-account-operator/pkg/localmetrics"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("AccountClaim", func() {
+
+	var (
+		nullLogger   testutils.NullLogger
+		name         = "testAccountClaim"
+		namespace    = "myAccountClaimNamespace"
+		accountClaim *awsv1alpha1.AccountClaim
+		r            *ReconcileAccountClaim
+		ctrl         *gomock.Controller
+	)
+
+	err := apis.AddToScheme(scheme.Scheme)
+	if err != nil {
+		fmt.Printf("failed adding apis to scheme in account controller tests")
+	}
+	localmetrics.Collector = localmetrics.NewMetricsCollector(nil)
+
+	BeforeEach(func() {
+		ctrl = gomock.NewController(GinkgoT())
+
+		accountClaim = &awsv1alpha1.AccountClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+			Spec: awsv1alpha1.AccountClaimSpec{
+				LegalEntity: awsv1alpha1.LegalEntity{
+					Name: "LegalCorp. Inc.",
+					ID:   "abcdefg123456",
+				},
+				AccountLink: "osd-creds-mgmt-aaabbb",
+				Aws: awsv1alpha1.Aws{
+					Regions: []awsv1alpha1.AwsRegions{
+						{
+							Name: "us-east-1",
+						},
+					},
+				},
+			},
+		}
+
+		// Create the reconciler with a mocking AWS client IBuilder.
+		r = &ReconcileAccountClaim{
+			// Test cases need to set fakeClient.
+			scheme: scheme.Scheme,
+			awsClientBuilder: &mock.Builder{
+				MockController: ctrl,
+			},
+		}
+	})
+
+	AfterEach(func() {
+		ctrl.Finish()
+	})
+
+	Context("Finalizers", func() {
+		It("should add finalizer correctly", func() {
+			// Objects to track in the fake client.
+			objs := []runtime.Object{accountClaim}
+			r.client = fake.NewFakeClient(objs...)
+
+			err := r.addFinalizer(nullLogger, accountClaim)
+			Expect(err).NotTo(HaveOccurred())
+		})
+		It("should not add finalizer as account claim doesn't exist", func() {
+			// Objects to track in the fake client.
+			objs := []runtime.Object{}
+			r.client = fake.NewFakeClient(objs...)
+
+			err := r.addFinalizer(nullLogger, accountClaim)
```

Comment:
Would have liked to add validation on the accountClaim finalizer list itself, to ensure the size actually changes. But given these funcs manipulate the accClaim pointer and even if they encounter an error while making a request to k8s, that accClaim object will still have the modified finalizer list. Could potentially work around this by pulling the accClaim object in the controller, but seems excessive - but open to discussion... if any of this made sense

Reply:
If you run a separate "get" on NewFakeClient at the end of the test, that should only return what's "saved" in k8s I believe.

So the test would be something like the pseudocode below:

```
client := fake.NewFakeClient(objs...)
r.client = client

r.addFinalizer(nullLogger, accountClaim)
Expect(err).To(HaveOccurred())

ac := &awsv1alpha1.AccountClaim{}
testErr := client.Get(context.TODO(), ac)

Expect(TestErr).NotTo(HaveOccurred())
Expect(ac.ObjectMeta.Finalizers).To(BeEmpty())
```